src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -2164,6 +2167,13 @@ var projectUniqueNamesForBuildIntToUpdate
                         }
                     }
 
+                    if (msbuildProject != null)
+                    {
+                        // end batch for msbuild and let it save everything.
+                        // always calls it before PostProcessAsync or binding redirects
+                        msbuildProject.ProjectSystem.EndProcessing();
```

Comment:
Agree with @dtivel. For instance, here `BatchStart/BatchEnd` pair follows the try-finally pattern.